File: DSAlgorithms/patterns/2_pointers/8_dutch_flag.cpp
#include <bits/stdc++.h>
class dutch_flag {
    public:
        static void sort(std::vector<int> &arr)
        {
            int low = 0;
            int high = arr.size() - 1;
            for(int i = 0; i <= high;)
            {
                if(arr[i] == 0)
                {
                    swap(arr, i, low);
                    i++;
                    low++;
                }
                else if(arr[i] == 1)
                    i++;
                else
                {
                    swap(arr, i, high);
                    high--;
                }
            }
        }
    private: 
        static void swap(std::vector<int> &arr, int i, int j)
        {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
};

int main(int argc, char *argv[])
{
    std::vector<int> arr = {1, 0, 2, 1, 0};
    dutch_flag::sort(arr);
    for(auto num : arr)
        std::cout<<num<<" ";
    std::cout<<'\n';

    arr = {2, 2, 0, 1, 2, 0};
    dutch_flag::sort(arr);
    for(auto num : arr)
        std::cout<<num<<" ";
    std::cout<<'\n';
}

File: dsa_algo/patterns/2_pointers/8_dutch_flag.ts
const dutch_flag_sort = (arr : number[]) => {
    let low = 0;
    let high = arr.length - 1;
    let i = 0;
    while(i <= high)
    {
        if(arr[i] === 0)
        {
            [arr[i], arr[low]] = [arr[low], arr[i]];  //swap
            i++;
            low++;
        }
        else if(arr[i] === 1)
            i++;
        else
        {
            [arr[i], arr[high]] = [arr[high], arr[i]];  //swap
            high--;
        }
    }
}

let arr = [1, 0, 2, 1, 0];
dutch_flag_sort(arr);
console.log(arr);

arr = [2, 2, 0, 1, 2, 0];
dutch_flag_sort(arr);
console.log(arr);
